src/firefly/java/edu/caltech/ipac/firefly/server/packagedata/DownloadScriptWorker.java

```diff
@@ -0,0 +1,221 @@
+                    DownloadScript.composeDownloadScript(outFile, "Euclid", urlList, attribute);
+
+                    if (outFile.exists()) {
+                        String fileExtension = scriptTypeToExtension.getOrDefault(scriptType.toLowerCase(), ".txt");
+                        getJob().addResult(new JobInfo.Result(makeDownloadUrl(outFile, suggestedName + fileExtension), null, MediaType.PLAIN_TEXT_UTF_8.toString(), outFile.length() + ""));
+
+                        // handle 'save to Workspace' option:  pushes downloaded script (.sh or .txt) to workspace
+                        if (!isEmpty(wsDestPath)) {
+                            String wsFilePath = wsDestPath + suggestedName + fileExtension;
+                            try {
+                                new WsServerUtils().putFile(
+                                        new WsServerParams().set(CURRENTRELPATH, wsFilePath),
+                                        outFile);
+                            } catch (IOException e) {
+                                logger.error(e);
+                            }
+                        }
+                    }
+                }
+            }
+            catch(Exception e){
+                failed.add(e.getMessage());
+            }
+            hasErrors = hasErrors || !failed.isEmpty();
+        }
+        else {
+            throw new DataAccessException("Operation aborted:" + dlreq.getRequestId(), new IllegalArgumentException("Unable to resolve a search processor for this request"));
+        }
+
+        // JobInfo completion update
+        String summary = String.format("%,d files were processed.", totalFiles);
+        if (hasErrors) summary += "\nPlease, note:  There were error(s) while processing your request.";
+        JobInfo jobInfo = getJob().getJobInfo();
+        jobInfo.setProgress(100);
+        jobInfo.setProgressDesc(summary);
+        jobInfo.setSummary(summary);
+        getJob().setPhase(JobInfo.Phase.COMPLETED);
+
+        PackagedEmail.send(getJob().getJobInfo());
+
+        return "";
+    }
+
+    private void updateJobProgress() throws DataAccessException.Aborted {
+        Job job = getJob();
+        if (job != null) {
+            JobInfo.Phase phase = job.getJobInfo().getPhase();
+            if (phase == JobInfo.Phase.ABORTED) throw new DataAccessException.Aborted();
+            if (System.currentTimeMillis() - lastUpdatedTime > 2000) {
+                lastUpdatedTime = System.currentTimeMillis();
+                int pct = (int) ((float)curFileInfoIdx / totalFiles * 100);
+                if ( pct != lastUpdatedPct) {
+                    lastUpdatedPct = pct;
+                    job.progress(pct, String.format("%d of %d completed", curFileInfoIdx, totalFiles));
+                }
+            }
+        }
+    }
+
+    private static String makeDownloadUrl(File f, String suggestedName) {
+        String fileStr = ServerContext.replaceWithPrefix(f);
```

Comment:
No need to copy/paste this function here.  Change PackagingWorker.makeDownloadUrl accessor to public or package protected(remove private) then use it instead.